pkg/cmd/pulumi/plugin_install.go

```diff
@@ -62,161 +59,204 @@ func newPluginInstallCmd() *cobra.Command {
+}
+
+func (cmd *pluginInstallCmd) Run(ctx context.Context, args []string) error {
+	if cmd.e == nil {
+		cmd.e = env.Global()
+	}
+	if cmd.d == nil {
+		cmd.d = cmdutil.Diag()
+	}
+	if cmd.color == "" {
+		cmd.color = cmdutil.GetGlobalColorization()
+	}
+
+	// Parse the kind, name, and version, if specified.
+	var installs []workspace.PluginSpec
+	if len(args) > 0 {
+		if !workspace.IsPluginKind(args[0]) {
+			return fmt.Errorf("unrecognized plugin kind: %s", args[0])
+		} else if len(args) < 2 {
+			return errors.New("missing plugin name argument")
+		}
+
+		var version *semver.Version
+		if len(args) == 3 {
+			parsedVersion, err := semver.ParseTolerant(args[2])
+			version = &parsedVersion
+			if err != nil {
+				return fmt.Errorf("invalid plugin semver: %w", err)
 			}
+		}
+		if len(args) < 3 && cmd.file != "" {
+			return errors.New("missing plugin version argument, this is required if installing from a file")
+		}
 
-			// Parse the kind, name, and version, if specified.
-			var installs []workspace.PluginSpec
-			if len(args) > 0 {
-				if !workspace.IsPluginKind(args[0]) {
-					return fmt.Errorf("unrecognized plugin kind: %s", args[0])
-				} else if len(args) < 2 {
-					return errors.New("missing plugin name argument")
-				}
+		var checksums map[string][]byte
+		if cmd.checksum != "" {
+			checksumBytes, err := hex.DecodeString(cmd.checksum)
+			if err != nil {
+				return fmt.Errorf("--checksum was not a valid hex string: %w", err)
+			}
+			checksums = map[string][]byte{
+				runtime.GOOS + "-" + runtime.GOARCH: checksumBytes,
+			}
+		}
 
-				var version *semver.Version
-				if len(args) == 3 {
-					parsedVersion, err := semver.ParseTolerant(args[2])
-					version = &parsedVersion
-					if err != nil {
-						return fmt.Errorf("invalid plugin semver: %w", err)
-					}
-				}
-				if len(args) < 3 && file != "" {
-					return errors.New("missing plugin version argument, this is required if installing from a file")
-				}
+		pluginSpec := workspace.PluginSpec{
+			Kind:              workspace.PluginKind(args[0]),
+			Name:              args[1],
+			Version:           version,
+			PluginDownloadURL: cmd.serverURL, // If empty, will use default plugin source.
+			Checksums:         checksums,
+		}
 
-				var checksums map[string][]byte
-				if checksum != "" {
-					checksumBytes, err := hex.DecodeString(checksum)
-					if err != nil {
-						return fmt.Errorf("--checksum was not a valid hex string: %w", err)
-					}
-					checksums = map[string][]byte{
-						runtime.GOOS + "-" + runtime.GOARCH: checksumBytes,
-					}
-				}
+		// Bundled plugins are generally not installable with this command. They are expected to be
+		// distributed with Pulumi itself. But we turn this check off if PULUMI_DEV is set so we can
+		// test installing plugins that are being moved to their own distribution (such as when we move
+		// pulumi-nodejs).
+		isBundled := (pluginSpec.Kind == workspace.LanguagePlugin && pluginSpec.Name == "nodejs") ||
```

Comment:
I'm very undecided about what's the right logic here.
1. Should it be we give this warning for all language plugins + the python/node policy/resource, on the basis that we don't _really_ support language as plugins yet, they are tightly coupled to the same CLI version (we _do_ make grpc breaking changes to the language interface).
2. Should it error for all our plugins, but if someone tries to install a community rust (for example) language plugin we just allow that (even if DEV=false)
3. Should it error for just the plugins which we know can't install via plugin install, i.e. nodejs, go, and python language plugins + the nodejs/python policy/resource. So if someone _wants_ to install dotnet like this they can, even though it's not really supported.

I'm currently leaning towards 1 or 2.

Reply:
(2) sounds more reasonable to me. 
We should only consider what we know is bundled with pulumi as bundled for the purposes of this command. This avoids that assumption about what's supported by the wider tooling.
So if/when we do support other language plugins, nobody should have to come back here and fix this.

Reply:
OK I've reworked this to explicitly list _our_ plugins and only our plugins as "bundled". Any other language plugin will not look next to the pulumi binary, and will be allowed to install via "plugin install"